### ISSUE TITLE
PIMD not working with FreeBSD12.2 if MAXVIFS-64 (I created a FreeBSD-ticket) 

### ISSUE THREAD

**Issue by LouisAtGH** [troglobit/pimd#180]
Joachim,

Lot of thanks for the support!, not the hoped result, but shit happens.
At least the conclusion was important "very likely FreeBSD-issue".

Saturday I did not realize that they already performed that maxvifs change, I should have known. Perhaps we would have done different tests.  Sorry.

I did more testing after our session, also adding debug statements. That did, as expected, not change the FreeBSD-issue-conclusion. To be even more sure, and to have a working system, I did contacted NetGate and they did roll-back the MAXVIFS change.

Retesting learned that that maxvifs back to 32 ^solved^ :) the issue. 

After all these tests, all hinting in the direction of a "FreeBSD maxvifs-problem", I did open a FreeBSD bug report (251047 Multicast: Not possible to add MFC entry's when MAXVIFS = 64).

I will keep you informed if there is progress.

Louis

**Comment by LouisAtGH** [comment]
Problem solved,

I made a mistake. The main problem was that expected that changing MAXVIFS at one place (in the program), would do the job. Far from that.

The following changes where/are required:
-	Updating the include files within the repository (include/freebsd/netinet) with the ones supplied with FreebSD 12.2-FINAL (/usr/include/netinet)
-	Changing the MAXVIFS settings in  ip_mroute.h to 64 (within that netinet folder)
-	Changing the MAXVIFS settings from the development machine itself, being in  /usr/include/netinet
-	Changing the “configure option in the build script to
    ./configure --localstatedir=/var  --with-max-vifs=64
-	**Open issue-1**: Changing the FreeBSD ports build script to use maxvifs64. For that I used a workaround (see below)

For FreeBSD you need to / normally build the package in two steps:
Step-1: building the package with GIT/GitHub that result in a “distribution file”
Step-2: building the package with FreebSD ports using the “distribution file”
After the second step you have the package to install.


**Work around Issue-1 option-1** (the used solution)
Since the configure “max-vifs” option does not reach FreeBSD-ports, ports is building the package using the default maxvifs32.  That is despite the fact that I changed the machine its maxvifs settings in ip_mroute.h. So the generated package has maxvifs32 and not the maxvifs64 we need.
To work around that:
-	I installed the “max-vifs32” package” and then
-	Replaced the pimd and pimctl files, with the ones generated with GIT.
That seems to work 

**Work around issue-1 option-2** (perhaps, not used)
could perhaps be changing the XX in following configure.ac line from 32 to 64. 
as a work around did use configure ac 
AS_IF([test "x$with_max_vifs" != "xno" -a "x$max_vifs" != "xyes"], [AC_DEFINE_UNQUOTED(CUSTOM_MAX_VIFS, $max_vifs, [Custom MAX VIFs in kernel.])],
[ max_vifs=^XX^])
Advantage is that that change does reach ports. However I did choose not to use that option, since I am not sure about impact of that change.


**Open issue-2**
As said I used exactly the includes as provided by FreeBSD. That to be sure that differences would not be the cause of the problem. However …….. those files significantly differ from the ones included in the repository. Especially the in.h file is different.  In the ip_mroute.h one of the differences is the use of uint64_t mrts <>   u_long (I do not know if a long differs from a 64bit integer!?)
What ever the compiler did not complain.

General impression was that there where extra lines in in.h, from thinks (not yet) supported by actual FreeBSD-version.

So bottom line, issue solved apart from the mentioned open issues.


Louis

**Comment by troglobit** [comment]
Bug for unsupported operating system, closing.
